Blammo-wai/src/Network/Wai/Middleware/Logging.hs

```diff
@@ -144,10 +144,39 @@ requestLoggerWith config env app req respond =
     app req $ \resp -> do
       recvd <- respond resp
       duration <- toMillis . subtract begin <$> getTime
-      recvd <$ runInIO (runWithLogger env $ logResponse config duration req resp)
+      runInIO $
+        runWithLogger env $
+          if isRaw resp
+            then logRawResponse config duration req
+            else logResponse config duration req resp
+      pure recvd
  where
   getTime = Clock.getTime Clock.Monotonic
   toMillis x = fromIntegral (Clock.toNanoSecs x) / nsPerMs
+  isRaw = \case
+    ResponseRaw {} -> True
+    _ -> False
+
+logRawResponse :: MonadLogger m => Config -> Double -> Request -> m ()
+logRawResponse Config {..} duration req =
+  logDebugNS cLogSource $ message :# details
+ where
+  message =
+    decodeUtf8 (requestMethod req)
+      <> " "
+      <> decodeUtf8 (rawPathInfo req)
```

Comment:
How about replacing `decodeUtf8` with `\x -> either (T.pack $ show x) id $ decodeUtf8' x` so that non-UTF8 oddities can display their oddities instead of throwing?

Reply:
I believe that's `decodeUtf8With lenientDecode` with more steps.

I'm cool with that, but as this code was existing, I think I'd like to do it separately. Maybe I'll file an Issue so we don't forget.

Reply:
Oh, if it's not valid UTF-8, will `show` actually give you something better than the box?

Seems like up-streaming a way to do `decodeUtf8With showingDecode` or something would be valuable then.

Reply:
Yeah, IMO `show` on `ByteString` is pretty good for situations where you have a byte string that you hoped would be text but might be anything. For each byte it prints either the ASCII interpretation or backslash escapes, so basic text comes out readable, and non-ascii junk is ugly but at least visible.

```haskell
ghci> print $ pack [65..90]
"ABCDEFGHIJKLMNOPQRSTUVWXYZ"

ghci> print $ pack [120..130]
"xyz{|}~\DEL\128\129\130"

```

Reply:
Ah fair enough. I dunno that it's really _that_ much better...

```console
λ> decodeUtf8With lenientDecode $ BS.pack [120..130]
"xyz{|}~\DEL\65533\65533\65533"
λ> print $ BS.pack [120..130]
"xyz{|}~\DEL\128\129\130"
```

128, 129, 130 are really no more meaningful to me than calling them all 65533, but I guess it's "better" not to be lossy.